### PR TITLE
Fix #2613 - Add linter for non-existing vars in refer-clojure

### LINKED ISSUE #2613: Warn on non-existing :refer-clojure :exclude var

### PR DESCRIPTION
**Pull request by jramosg**
# Fix #2613
Introduce a new linter `:refer-clojure-exclude-non-existing-var` to warn users about non-existing vars specified in the `:exclude` option of `:refer-clojure`. This enhancement improves code quality by preventing potential runtime errors due to incorrect references.

Please answer the following questions and leave the below in as part of your PR.

- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

- [x] I have read the [developer documentation](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md).

- [x] This PR corresponds to an [issue with a clear problem statement](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).

- [x] This PR contains a [test](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#tests) to prevent against future regressions

- [x] I have updated the [CHANGELOG.md](https://github.com/clj-kondo/clj-kondo/blob/master/CHANGELOG.md) file with a description of the addressed issue.

### CODE REVIEW

**Comment by jramosg** [comment]
Hello @borkdude 

Why `def` is not in `clojure-core-syms` and `cljs-core-syms`? Is there any reason?

Thank you!

**Comment by borkdude** [comment]
@jramosg 

Hey, thanks for this PR! Is it possible to stop using force-push when updating PRs?

See: https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#force-push

To answer your question:

```
def
```

isn't a var but a special form.

See e.g.:

```
(and call?
                  (special-symbol? name-sym))
```

in `src/clj_kondo/impl/namespace.clj`

**Comment by borkdude** [comment]
It seems that excluding `def` doesn't do anything.

```
user=> (ns foo (:refer-clojure :exclude [def]))
nil
foo=> (def x 1)
#'foo/x
```

Where did you see that?

File: src/clj_kondo/impl/var_info.clj [hunk 1]
```diff
@@ -5,7 +5,7 @@
 (declare clojure-core-syms cljs-core-syms)
 
 ;; in addition to what `special-form?` regards as special:
-(def special-forms '#{.. let fn loop})
+(def special-forms '#{.. let fn loop def if recur var do quote})
```
Comment (borkdude) [inline src/clj_kondo/impl/var_info.clj]: Please read the comment above this form: in _addition_ to what `special-form?` regards as special. `def` etc are already covered by `special-form?`.

File: src/clj_kondo/impl/analyzer/namespace.clj [hunk 2]
```diff
@@ -463,6 +463,37 @@
...
+           (node->line
+            filename excluded-var
+            :refer-clojure-exclude-non-existing-var
+            (format "The var %s does not exist in %s"
+                    excluded-var
+                    (case lang
+                      :clj "clojure.core"
+                      :cljs "cljs.core"
+                      :cljc (str/join "and"
+                                      (cond-> []
```
Comment (jramosg) [inline src/clj_kondo/impl/analyzer/namespace.clj]: This cljc check is useless, I will delete it

File: src/clj_kondo/impl/analyzer/namespace.clj [hunk 3]
```diff
@@ -463,6 +463,26 @@
...
 
+(defn- lint-refer-clojure-vars [{:keys [filename lang] :as ctx} excluded-vars]
+  (letfn [(exists-in-core?  [excluded-var lang]
+            (if (= :cljc (:base-lang ctx))
+              (some #(core-sym? % excluded-var) [:clj :cljs])
+              (core-sym? lang excluded-var)))]
+    (when-not (linter-disabled? ctx :refer-clojure-exclude-non-existing-var)
+      (doseq [excluded-var excluded-vars
+              :when (not (or (special-symbol? excluded-var)
+                             (contains? special-forms excluded-var)
```
Comment (borkdude) [inline src/clj_kondo/impl/analyzer/namespace.clj]: This line does not appear to be necessary since `..`, `let`, etc already are clojure vars and thus are covered by `exists-in-core?`.

File: src/clj_kondo/impl/analyzer/namespace.clj [hunk 4]
```diff
@@ -463,6 +463,26 @@
...
    :col col})
 
+(defn- lint-refer-clojure-vars [{:keys [filename lang] :as ctx} excluded-vars]
+  (letfn [(exists-in-core?  [excluded-var lang]
+            (if (= :cljc (:base-lang ctx))
+              (some #(core-sym? % excluded-var) [:clj :cljs])
+              (core-sym? lang excluded-var)))]
+    (when-not (linter-disabled? ctx :refer-clojure-exclude-non-existing-var)
+      (doseq [excluded-var excluded-vars
+              :when (not (or (special-symbol? excluded-var)
```
Comment (borkdude) [inline src/clj_kondo/impl/analyzer/namespace.clj]: This line also doesn't appear to be necessary. When I comment it out and run `lein do clean, test :only clj-kondo.refer-clojure-test` all tests still pass, so it's not even covered by the tests.
You can't exclude a special form in Clojure, so let's remove it altogether.

Comment (borkdude) [reply src/clj_kondo/impl/analyzer/namespace.clj]: So in summary: the only thing that matters is: does it exist in core

File: test/clj_kondo/refer_clojure_test.clj [hunk 5]
```diff
@@ -0,0 +1,120 @@
+(ns clj-kondo.refer-clojure-test
+  (:require
+   [clj-kondo.test-utils :refer [assert-submaps lint!]]
```
Comment (borkdude) [inline test/clj_kondo/refer_clojure_test.clj]: Please use assert-submaps2 from now it, it has better error reporting.

File: src/clj_kondo/impl/analyzer/namespace.clj [hunk 6]
```diff
@@ -463,6 +463,24 @@
...
    :col col})
 
+(defn- lint-refer-clojure-vars [{:keys [filename lang] :as ctx} excluded-vars]
+  (letfn [(exists-in-core?  [excluded-var lang]
+            (if (= :cljc (:base-lang ctx))
+              (some #(core-sym? % excluded-var) [:clj :cljs])
+              (core-sym? lang excluded-var)))]
+    (when-not (linter-disabled? ctx :refer-clojure-exclude-non-existing-var)
+      (doseq [excluded-var excluded-vars
+              :when (exists-in-core? excluded-var lang)]
```
Comment (borkdude) [inline src/clj_kondo/impl/analyzer/namespace.clj]: I think you forgot a negation here :)

**Comment by jramosg** [comment]
Hello @borkdude 

Regarding to this comments (https://github.com/clj-kondo/clj-kondo/pull/2675#discussion_r2637024252 and https://github.com/clj-kondo/clj-kondo/pull/2675#discussion_r2637025292) `(clj-kondo.impl.var-info/core-sym? lang excluded-var)` I updated the new tests to correcty ensure them and also I had to update some existing tests

**Comment by borkdude** [comment]
@jramosg Last remark. Since this isn't a very critical linter and it's a new one, we could put the level to `:info` initially. Perhaps people will find some edge cases and when putting it to `:info`, it won't break their CI.
Another consideration about the name: 

```
:refer-clojure-exclude-non-existing-var
```

In clj-kondo's existing naming style, we call a non-existing var also a `unresolved-var`. So perhaps

```
:refer-clojure-exclude-unresolved-var
```

maybe?

**Comment by jramosg** [comment]
@borkdude changed, thank you!
